### PR TITLE
Slamd 132

### PR DESCRIPTION
**Pull request by iteratec-hscaffino**

### CODE REVIEW

File: slamd/discovery/processing/experiment/models/tuned_gaussian_process_regressor.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                                      scoring='r2',
+                                      n_jobs=-1,
+                                      cv=4,
+                                      refit=False)
+        grid_search_cv = grid_search_cv.fit(training_rows, training_labels)
+        # Return the best model found
+        return pipe.set_params(**grid_search_cv.best_params_)
+
+    @classmethod
+    def _create_optimized_gaussian_process(cls):
```
Comment (RugTehseen) [inline slamd/discovery/processing/experiment/models/tuned_gaussian_process_regressor.py:32]: maybe we can find a better name. This Gauss is not yet optimized, but the chosen fixed parameters rather serves as a sensible starting point for the subsequent tuning. Maybe somethin like _create_baseline_gauss_for_further_tuning?

File: slamd/discovery/processing/experiment/models/tuned_gaussian_process_regressor.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+from mlxtend.feature_selection import SequentialFeatureSelector
```
Comment (RugTehseen) [inline slamd/discovery/processing/experiment/models/tuned_gaussian_process_regressor.py:1]: the package containing the ml models is now called models. I would prefer someting like mlmodel. This way it is clear that these classes are not models in the way we usually use models, i.e. as session objects.

File: slamd/discovery/processing/experiment/models/tuned_gaussian_process_regressor.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    @classmethod
+    def _create_anisotropic_kernel(cls, n_dims):
+        return ConstantKernel(1.0, constant_value_bounds='fixed') * RBF(length_scale=[1] * n_dims)
+
+    @classmethod
+    def _create_parameters_for_grid_search(cls):
+        """
+        Return the grid parameters required by GridSearchCV.
+        Use a isotropic and an anisotropic kernel.
```
Comment (RugTehseen) [inline slamd/discovery/processing/experiment/models/tuned_gaussian_process_regressor.py:48]: "Use an isotropic" instead of "Use a isotropic"

File: slamd/discovery/processing/experiment/models/tuned_random_forest.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+from mlxtend.feature_selection import SequentialFeatureSelector
+from sklearn.model_selection import GridSearchCV
+from sklearn.pipeline import Pipeline
+
+
+class TunedRandomForest:
+
+    @classmethod
+    def find_best_model(cls, training_rows, training_labels):
+        # Perform a forward sequential feature selection with 3 features without cross-validation
```
Comment (RugTehseen) [inline slamd/discovery/processing/experiment/models/tuned_random_forest.py:12]: Here and for Tuned Gauss. We actually never perform a seq. feat. selection with 3 features. It is overwritten by the values in our grid which are 5 and 10, respectively. I guess we could init we the default value here. It is just an artefact that we have a 3 here. Please check if you can use the default and it still works.

File: slamd/discovery/processing/experiment/models/tuned_gaussian_process_regressor.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                                                   floating=False,
+                                                   scoring='r2',
+                                                   cv=None)
+        pipe = Pipeline([('sfs', sfs_gr_testing),
+                         ('gp2', cls._create_optimized_gaussian_process())])
+        grid_search_cv = GridSearchCV(estimator=pipe,
+                                      param_grid=cls._create_parameters_for_grid_search(),
+                                      scoring='r2',
+                                      n_jobs=-1,
+                                      cv=4,
```
Comment (RugTehseen) [inline slamd/discovery/processing/experiment/models/tuned_gaussian_process_regressor.py:25]: Here and below for RF. We need to be careful with the cv param. In case we have a dataset with less than 4 labels for the target this would crash. I would simply add in the infobox when running the app locally something like: "The tuned versions require the targets to have at least 4 labels". You should also validate this in the backend and raise a custom exceeption if number_labels < 4

File: slamd/discovery/processing/experiment/experiment_model.py [hunk 6]
```diff
@@ -1,6 +1,22 @@
...
     GAUSSIAN_PROCESS = 'Statistics-based model (Gaussian Process Regression)'
+    RANDOM_FOREST = 'AI Model (lolo Random Forest)'
+    PCA_GAUSSIAN_PROCESS = 'Gaussian Process with PCA'
+    PCA_RANDOM_FOREST = 'Random Forest with PCA'
+    TUNED_GAUSSIAN_PROCESS = 'Tuned Gaussian Process'
+    TUNED_RANDOM_FOREST = 'Tuned Random Forest'
+
+    @classmethod
+    def get_all_models(cls):
+        if os.getenv('FLASK_ENV') != 'demo':
```
Comment (RugTehseen) [inline slamd/discovery/processing/experiment/experiment_model.py:15]: This works, but it might be better to turn the logic around and say if os.getenv('FLASK_ENV') == 'development'. This way, also when Christoph adds a new profile / heroku instance e.g. 'production' it is still only enabled locally and not on prod. The way it is implemented right now would lead to the effect that for a new instance / environment the tuned versions are automatically enabled even though we always only want them locally.

**Review by RugTehseen — APPROVED** [review]
looks good now
